From da93f3c37a54f6b8e2ad72e0b1a255f3257adc9f Mon Sep 17 00:00:00 2001
From: William Wu <william.wu@rock-chips.com>
Date: Thu, 19 Nov 2020 11:23:44 +0800
Subject: [PATCH] phy: rockchip: naneng-combphy: enable adaptive CTLE for
 RK3568 USB3.0

This patch enable the adaptive Continuous Time Linear Equalizer (CTLE)
for RK3568 USB3.0 to improve compatibility.

Change-Id: I04d3077e37f15a8d41df875b8d84dc7e6c8aeda9
Signed-off-by: William Wu <william.wu@rock-chips.com>
---
 .../rockchip/phy-rockchip-naneng-combphy.c    | 25 ++++++++++++++++---
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/drivers/phy/rockchip/phy-rockchip-naneng-combphy.c b/drivers/phy/rockchip/phy-rockchip-naneng-combphy.c
index 502b71fa5928..eb68977bfc1d 100644
--- a/drivers/phy/rockchip/phy-rockchip-naneng-combphy.c
+++ b/drivers/phy/rockchip/phy-rockchip-naneng-combphy.c
@@ -317,8 +317,15 @@ static int rk3568_combphy_cfg(struct rockchip_combphy_priv *priv)
 	case 24000000:
 		if (priv->mode == PHY_TYPE_USB3) {
 			/* Set ssc_cnt[9:0]=0101111101 & 31.5KHz */
-			writel(0x40, priv->mmio + 0x38);
-			writel(0x5f, priv->mmio + 0x3c);
+			val = readl(priv->mmio + (0x0e << 2));
+			val &= ~GENMASK(7, 6);
+			val |= 0x01 << 6;
+			writel(val, priv->mmio + (0x0e << 2));
+
+			val = readl(priv->mmio + (0x0f << 2));
+			val &= ~GENMASK(7, 0);
+			val |= 0x5f;
+			writel(val, priv->mmio + (0x0f << 2));
 		}
 		break;
 	case 25000000:
@@ -350,8 +357,18 @@ static int rk3568_combphy_cfg(struct rockchip_combphy_priv *priv)
 		writel(val, priv->mmio + (0x20 << 2));
 		break;
 	case PHY_TYPE_USB3:
-		/* Set ssc downward spread spectrum */
-		writel(0x10, priv->mmio + 0x7c);
+		/* Set SSC downward spread spectrum */
+		val = readl(priv->mmio + (0x1f << 2));
+		val &= ~GENMASK(5, 4);
+		val |= 0x01 << 4;
+		writel(val, priv->mmio + 0x7c);
+
+		/* Enable adaptive CTLE for USB3.0 Rx */
+		val = readl(priv->mmio + (0x0e << 2));
+		val &= ~GENMASK(0, 0);
+		val |= 0x01;
+		writel(val, priv->mmio + (0x0e << 2));
+
 		param_write(priv->phy_grf, &cfg->pipe_sel_usb, true);
 		param_write(priv->phy_grf, &cfg->pipe_txcomp_sel, false);
 		param_write(priv->phy_grf, &cfg->pipe_txelec_sel, false);
-- 
2.35.3

